File: jvm_app/src/main/kotlin/io/neverdo/jvm/ui/ListFormView.kt
package io.neverdo.jvm.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.neverdo.shared.db.BoardDb
import io.neverdo.shared.db.ListDb
import io.neverdo.shared.vm.ListFormVM

@Composable
fun ListFormView(
    list: ListDb?,
    board: BoardDb,
    modifier: Modifier,
    onCancel: () -> Unit,
    onSave: () -> Unit,
) {

    val (vm, state) = rememberVM(list) { ListFormVM(list, board) }

    VStack(
        modifier = modifier,
    ) {

        MyTextField(
            value = state.inputText,
            onValueChange = {
                vm.setInputText(it)
            },
            modifier = Modifier
                .fillMaxWidth(),
            placeholder = "List Name",
        )

        HStack(
            modifier = Modifier
                .padding(top = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {

            Text(
                text = "Save",
                modifier = Modifier
                    .clip(squircleShape)
                    .background(blueColor)
                    .clickable {
                        vm.submitForm {
                            onSave()
                        }
                    }
                    .padding(horizontal = 8.dp, vertical = 6.dp),
                color = Color.White,
                fontSize = 13.sp,
            )

            Text(
                text = "Cancel",
                modifier = Modifier
                    .clickable {
                        onCancel()
                    }
                    .padding(horizontal = 8.dp, vertical = 4.dp),
                fontSize = 13.sp,
                color = Color.Gray,
            )

            if (list != null)
                Text(
                    text = "Delete List",
                    modifier = Modifier
                        .padding(start = 8.dp)
                        .clip(squircleShape)
                        .background(Color.Red)
                        .clickable {
                            vm.deleteList()
                        }
                        .padding(horizontal = 8.dp, vertical = 6.dp),
                    color = Color.White,
                    fontSize = 13.sp,
                )
        }
    }
}

File: apple_apps/neverdo/ui/ListFormView.swift
import SwiftUI
import shared

struct ListFormView: View {

    @State private var vm: ListFormVM

    private let onCancel: () -> Void
    private let onSave: () -> Void

    ///

    @FocusState private var isFormFocused: Bool

    init(
        listDb: ListDb?,
        boardDb: BoardDb,
        onCancel: @escaping () -> Void,
        onSave: @escaping () -> Void
    ) {
        _vm = State(initialValue: ListFormVM(list: listDb, boardToAdd: boardDb))
        self.onCancel = onCancel
        self.onSave = onSave
    }

    var body: some View {

        VMView(vm: vm, stack: .VStack()) { state in

            ZStack {
                TextField__VMState(
                    text: state.inputText,
                    placeholder: "Board Name",
                    isFocused: $isFormFocused,
                    onValueChanged: { newValue in
                        vm.setInputText(text: newValue)
                    }
                )
            }
                    .background(squircleShape.fill(.tertiary))

            HStack(spacing: 10) {

                Button("Save") {
                    vm.submitForm {
                        onSave()
                    }
                }

                Button("Cancel") {
                    onCancel()
                }
                        .buttonStyle(.link)
                        .foregroundColor(.secondary)

                if state.list != nil {
                    Button("Delete List") {
                        vm.deleteList()
                    }
                }

                Spacer()
            }
                    .padding(.top, 16)
        }
                .frame(width: 250)
    }
}
